saturation.go

```diff
@@ -0,0 +1,148 @@
+package raft
+
+import (
+	"math"
+	"time"
+
+	"github.com/armon/go-metrics"
+)
+
+// saturationMetric measures the saturation (percentage of time spent working vs
+// waiting for work) of an event processing loop, such as runFSM. It reports the
+// saturation as a gauge metric (at most) once every reportInterval.
+//
+// Callers must instrument their loop with calls to sleeping and working, starting
+// with a call to sleeping.
+//
+// Note: it is expected that the caller is single-threaded and is not safe for
+// concurrent use by multiple goroutines.
+type saturationMetric struct {
+	reportInterval time.Duration
+
+	// buckets contains measurements for both the current, and a configurable
+	// number of previous periods (to smooth out spikes).
+	buckets       []saturationMetricBucket
+	currentBucket int
+
+	sleepBegan, workBegan time.Time
+
+	// These are overwritten in tests.
+	nowFn    func() time.Time
+	reportFn func(float32)
+}
+
+// saturationMetricBucket contains the measurements for a period.
+type saturationMetricBucket struct {
+	// beginning of the period for which this bucket contains measurements.
+	beginning time.Time
+
+	// slept contains time for which the event processing loop was sleeping rather
+	// than working.
+	slept time.Duration
+
+	// lost contains time that was lost due to incorrect instrumentation of the
+	// event processing loop (e.g. calling sleeping() or working() multiple times
+	// in succession).
```

Comment:
```suggestion
	// lost contains time that is considered lost due to incorrect use
	// of saturationMetricBucket (e.g. calling sleeping() or working() 
	// multiple times in succession).
```

Reply:
My reading of the code is that your functions _account_ for this time lost, but the original wording made it seem like the reporting would be wrong with incorrect use (correct me if I'm wrong!)